app/controllers/laboratories_controller.rb

```diff
@@ -2,34 +2,78 @@
 
 # LaboratoriesController: lab results
 class LaboratoriesController < ApplicationController
-  before_action :authenticate_user!
+  before_action :authenticate_user!, :check_role, :check_patient
 
   # Create a new lab result
   def create
-    redirect_to(root_url) && return unless current_user.can_create_patient_laboratories?
+    patient_id = params.permit(:patient_id)[:patient_id]&.to_i
 
     lab = Laboratory.new(lab_type: params.permit(:lab_type)[:lab_type],
                          specimen_collection: params.permit(:specimen_collection)[:specimen_collection],
                          report: params.permit(:report)[:report],
                          result: params.permit(:result)[:result])
     lab.patient_id = params.permit(:patient_id)[:patient_id]
     lab.save!
-    History.lab_result(patient: params.permit(:patient_id)[:patient_id],
+    History.lab_result(patient: patient_id,
                        created_by: current_user.email,
                        comment: "User added a new lab result (ID: #{lab.id}).")
   end
 
   # Update an existing lab result
   def update
-    redirect_to(root_url) && return unless current_user.can_edit_patient_laboratories?
+    patient_id = params.permit(:patient_id)[:patient_id]&.to_i
 
     lab = Laboratory.find_by(id: params.permit(:id)[:id])
     lab.update!(lab_type: params.permit(:lab_type)[:lab_type],
                 specimen_collection: params.permit(:specimen_collection)[:specimen_collection],
                 report: params.permit(:report)[:report],
                 result: params.permit(:result)[:result])
-    History.lab_result_edit(patient: params.permit(:patient_id)[:patient_id],
+
+    History.lab_result_edit(patient: patient_id,
                             created_by: current_user.email,
                             comment: "User edited a lab result (ID: #{lab.id}).")
   end
+
+  # Delete an existing lab result
+  def destroy
+    patient_id = params.permit(:patient_id)[:patient_id]&.to_i
+
+    lab = Laboratory.find_by(id: params.permit(:id)[:id])
+    lab.destroy
+    if lab.destroyed?
+      reason = params.permit(:delete_reason)[:delete_reason]
+      comment = "User deleted a lab result (ID: #{lab.id}"
+      comment += ", Type: #{lab.lab_type}" unless lab.lab_type.blank?
+      comment += ", Specimen Collected: #{lab.specimen_collection}" unless lab.specimen_collection.blank?
+      comment += ", Report: #{lab.report}" unless lab.report.blank?
+      comment += ", Result: #{lab.result}" unless lab.result.blank?
+      comment += "). Reason: #{reason}."
+      History.lab_result_edit(patient: patient_id,
+                              created_by: current_user.email,
+                              comment: comment)
+    else
+      render status: 500
+    end
+  end
+
+  private
+
+  def check_role
+    redirect_to(root_url) && return unless current_user.can_edit_patient_laboratories?
+  end
```

Comment:
We should account for the two possible cases here: `can_edit_patient_laboratories` should be used for `update` and `destroy` and `can_create_patient_laboratories` should be used for `create`. I realize those checks happen to be identical, but if one were to change, we'd want to see the behavior change automatically propagate to here.

Reply:
Good point